apptail/storage/tail_tracker.go

```diff
@@ -0,0 +1,173 @@
+package storage
+
+import (
+	"fmt"
+	"github.com/ActiveState/log"
+	"runtime"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type Tracker interface {
+	Update(instKey string, childKey string, childVal int64)
+	LoadTailers()
+	Remove(key string)
+	RegisterInstance(instKey string)
+	InitializeChildNode(instKey string, childkey string, offSet int64)
+	Submit()
+	StartSubmissionTimer(retentionPeriod time.Duration)
+	IsInstanceRegistered(instKey string) bool
+	IsChildNodeInitialized(instKey string, childkey string) bool
+	GetFileCachedOffset(instkey string, fname string) int64
+}
+
+type BoxedInt64 struct{ V int64 }
+
+type TailNode map[string]*BoxedInt64
+
+type Tailer struct {
+	Instances map[string]TailNode
+}
+
+type tracker struct {
+	storage       Storage
+	Cached        *Tailer // do not expose this, it should ONLY be updated via Tracker methods
+	mux           *sync.Mutex
+	timerStopChan chan struct{} // used to send quit signal to timer
+}
+
+var (
+	MinIOTicker = 5 * time.Second
+)
+
+func NewTracker(s Storage) Tracker {
+	return &tracker{
+		storage: s,
+		mux:     &sync.Mutex{},
+		Cached: &Tailer{
+			Instances: make(map[string]TailNode),
+		},
+		timerStopChan: make(chan struct{}),
+	}
+}
+
+func (t *tracker) StartSubmissionTimer(retentionPeriod time.Duration) {
+	if retentionPeriod.Seconds() <= MinIOTicker.Seconds() {
+		seconds := retentionPeriod / (1000 * time.Millisecond)
+		log.Warnf("IMPORTANT: Setting retention period to %ds will increase your IO Rate", seconds)
+
+	}
+	ticker := time.NewTicker(retentionPeriod)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				t.Submit()
+			case <-t.timerStopChan:
+				ticker.Stop()
+				return
+			}
+
+		}
+	}()
+}
+
+func (t *tracker) RegisterInstance(instKey string) {
+	t.mux.Lock()
+	if _, instance_exist := t.Cached.Instances[instKey]; !instance_exist {
+		t.Cached.Instances[instKey] = TailNode{}
+		t.formatMap("Current Status")
```

Comment:
I wonder if we'll hit issues down the road from dumping out the status with the lock held.  But then, it's not like we can do this without the lock held…